pandora-server-directory/src/utility.ts

```diff
@@ -2,3 +2,105 @@
+	private immediatePromise: Promise<void> | null = null;
+	private executingPromise: Promise<void> | null = null;
+	private shouldScheduleNext = false;
+
+	/**
+	 * Creates an instance of AsyncInterval.
+	 * @param task The asynchronous task to be executed at fixed intervals.
+	 * @param interval The interval in milliseconds between executions of the task.
+	 * @param errorHandler A callback function that handles errors thrown during the execution of the task.
+	 */
+	constructor(task: () => Promise<void>, interval: number, errorHandler: (error: unknown) => void) {
+		this.task = task;
+		this.interval = interval;
+		this.errorHandler = errorHandler;
+	}
+
+	/**
+	 * Starts the execution of the asynchronous task at the specified interval.
+	 * If the task is already running, calling this method has no effect.
+	 * @returns The instance of AsyncInterval for chaining.
+	 */
+	public start(): this {
+		if (this.intervalId == null) {
+			this.shouldScheduleNext = false;
+			this.intervalId = setInterval(() => this.runOnInterval(), this.interval).unref();
+			this.runOnInterval();
+		}
+		return this;
+	}
+
+	/**
+	 * Stops the execution of the asynchronous task.
+	 * If the task is not currently running, calling this method has no effect.
+	 * @returns The instance of AsyncInterval for chaining.
+	 */
+	public stop(): this {
+		if (this.intervalId != null) {
+			clearInterval(this.intervalId);
+			this.intervalId = null;
+		}
+		this.shouldScheduleNext = false;
+		return this;
+	}
+
+	/**
+	 * Executes the task immediately, outside of its regular interval.
+	 * If the task is currently executing, this method waits for the current execution to finish before starting a new execution.
+	 * This method must be awaited to know when the immediate execution has completed.
+	 * @returns A promise that resolves when the immediate execution of the task has completed.
+	 */
+	public immediate(): Promise<void> {
+		if (this.immediatePromise != null) {
+			return this.immediatePromise;
+		}
+		if (this.executingPromise == null) {
+			return this.createExecutionPromise();
+		}
+		const executingPromise = this.executingPromise;
+		this.immediatePromise = new Promise<void>((resolve, reject) => {
+			executingPromise
+				.then(() => {
+					this.immediatePromise = null;
+					this.createExecutionPromise().then(resolve, reject);
+				}, (error) => {
+					this.immediatePromise = null;
+					reject(error);
+				});
+		});
+		return this.immediatePromise;
+	}
+
+	private runOnInterval(): void {
+		if (this.executingPromise != null) {
+			this.shouldScheduleNext = true;
+			return;
+		}
+		this.createExecutionPromise()
+			.catch((error) => this.errorHandler(error));
+	}
+
+	private createExecutionPromise(): Promise<void> {
+		this.shouldScheduleNext = false;
+		this.executingPromise = this.task()
+			.finally(() => {
+				this.executingPromise = null;
+				if (this.shouldScheduleNext) {
+					this.runOnInterval();
+				}
+			});
```

Comment:
This is not enough to safely call `.task()` if it can potentially fail. If the `task` is not async method, but a function that returns promise, it can throw before the promise is returned - the `catch`/`finally` handlers won't be called in that case.
The simplest way to fix this is to make `createExecutionPromise` an `async` method that `await`s the `this.task()` call, wrapping it in a `try`/`finally` block.
This way you can also change the signature of `task` to be `Promise<void> | void` to allow reusing this for simpler timers as well.